pkg/controller/jobframework/reconciler.go

```diff
@@ -675,14 +676,27 @@ func (r *JobReconciler) getPodSetsInfoFromAdmission(ctx context.Context, w *kueu
 				return nil, err
 			}
 			for k, v := range flv.Spec.NodeLabels {
-				nodeSelector.NodeSelector[k] = v
+				podSetInfo.NodeSelectorOverwrite[k] = v
```

Comment:
why not just `NodeSelector`?

Reply:
`NodeSelectorOverwrite` is set based on admission, to preserve the behavior on master that the node selectors from admission overwrite what is in job nodeSelector (there is no error on conflict). I think we could simplify, and error on conflicts in that case, but it would mean changing the master behavior as a drive by. I prefer that to be done in a dedicated PR if we want. WDYT? FYI @mwielgus ?

Reply:
There shouldn't be any conflicts, because the node selectors coming from the job are used to filter out flavors. https://kueue.sigs.k8s.io/docs/concepts/resource_flavor/#resourceflavor-labels

Reply:
We have currently tests (unit) that show the overwrite behavior, and they are failing otherwise: https://github.com/kubernetes-sigs/kueue/blob/d099039f66b2e0c9d243791ab275912ac606e0b4/pkg/controller/jobs/job/job_controller_test.go#L201-L225. So, the "update" as a test case name suggests this is the intention.

Reply:
So, in this test, we have a job with "orig-key": "orig-val", and it gets overwritten during run.

Reply:
OTOH, I'm happy to adjust the test (drop the "update" case) in this PR if you think this is the way to go.

Reply:
Tentatively removed the idea of overwrite in https://github.com/kubernetes-sigs/kueue/pull/1180/commits/8555fd3124ac1234418bf9e5c3170ce4d4d2b51c. Updating description

Reply:
> There shouldn't be any conflicts, because the node selectors coming from the job are used to filter out flavors. https://kueue.sigs.k8s.io/docs/concepts/resource_flavor/#resourceflavor-labels

In almost situations, resourceFlavors are created by batch admins and jobs are created by batch users. So I guess that conflicts could happen. Certainly, when users set the proper parameters for flvs and jobs, conflicts won't happen.

Actually, current implementation, each integration controller overwrites podSpec's nodeSelector with podSetInfo one (flvs):

https://github.com/kubernetes-sigs/kueue/blob/c3cfc1827b05b65c0dbc8a56e1ee84e76810f9f1/pkg/controller/jobs/job/job_controller.go#L227-L228

However, I think the previous implementation (https://github.com/kubernetes-sigs/kueue/commit/8555fd3124ac1234418bf9e5c3170ce4d4d2b51c) looks a bit hacky.

```go
...
	newNodeSelector := make(map[string]string)
	for k, v := range o.NodeSelector {
		if _, exists := podSetInfo.NodeSelectorOverwrite[k]; !exists {
			newNodeSelector[k] = v
		}
	}
	if err := utilmaps.HaveConflict(podSetInfo.NodeSelector, newNodeSelector); err != nil {
...
```

In this implementation, maybe `utilmaps.HaveConflict(podSetInfo.NodeSelector, newNodeSelector)` always returns nil, right?

Reply:
>In almost situations, resourceFlavors are created by batch admins and jobs are created by batch users. So I guess that conflicts could happen. Certainly, when users set the proper parameters for flvs and jobs, conflicts won't happen.

 IIUC, if the job configured by user can only get assigned a flavor, if the `NodeLabels` match the job's nodeSelector. Thus we will not have conflicts later when starting the job. So, e2e `MergeKeepFirst` does not matter, and the unit tests only tested the overwrite in isolation, assuming a flavor with conflicting labels is assigned.

> In this implementation, maybe `utilmaps.HaveConflict(podSetInfo.NodeSelector, newNodeSelector)` always returns nil, right?

Not really, it is possible, that `NodeSelectorOverwrite` was empty. Then, `newNodeSelector = o.NodeSelector`, and this could conflict with `podSetInfo.NodeSelector`.

Reply:
/cc @trasc who added the [unit tests for node selector overwrite](https://github.com/kubernetes-sigs/kueue/blob/c3cfc1827b05b65c0dbc8a56e1ee84e76810f9f1/pkg/controller/jobs/job/job_controller_test.go#L201C4-L225). 

Did you have some e2e scenario in mind that could lead to the overwrite, or the purpose of the tests is to just document the unit / component behavior?

Reply:
It was added for good unit coverage. Indeed I don't think it should happen in real life.

Reply:
> It was added for good unit coverage. Indeed I don't think it should happen in real life.

I think we can leave an `update` test case since this is a unit test.